Assignment 4 - counting islands

ElizabethSvit opened the pull request.

Commit:
Assignment 4 - counting islands

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,73 @@
+    
+        tile_deque.pop_front();
+        
+        if (cur_tile.first - 1 >= 0 && table[cur_tile.first - 1][cur_tile.second]) {
```
A loop on the 4 cardinal directions will be good here (either using two arrays for delta_x & delta_y, or looping over all 8 directions and only using the ones you need).

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,73 @@
+            table[cur_tile.first][cur_tile.second - 1] = false;
+            tile_deque.push_back({cur_tile.first, cur_tile.second - 1});
+        }
+        if (cur_tile.second + 1 < table[0].size() && table[cur_tile.first][cur_tile.second + 1]) {
```
Not a big deal, but what will happen if the first row on the table is bigger than the rest?

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,73 @@
+    flood_fill_recursive(table, y, x + 1);
+}
+
+// Second variant of algorithm using data structure deque for storing islands
```
Good that you have both variants, when will you use each? :)

ElizabethSvit replied:
It was just for practising both recursive and non-recursive algorithms :) I left only recursive for now. Also added the second algorithm Union-Find from Hint 2.

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,73 @@
+
+// First variant of algorithm using recursion
+
+void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
```
A pointer is better as a parameter you want to modify, make the calling code more explicit, compare:
`flood_fill_recursive(table, y, x)
`vs
`flood_fill_recursive(&table, y, x)`
In the second is obvious that the table will be modified.

ssancho2 reviewed (commented):
Looks good! Did you think of the other solution (re: hint 2 on the assignment) ?

Commit:
Updated assignment 4 with second algoritm

Left only recursive DFS algorithm from previous version and updated code with review comments being considered. Added the second Union-Find algorithm using Hint 2

ssancho2 commented on ass4.cpp:
```diff
@@ -1,12 +1,11 @@
+// Recursive "Flood fill" algorithm
 
-void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
+void flood_fill_recursive(vector<vector<bool> > table, int y, int x) {
```
This way you are making a copy of the whole table for every call, and you are not modifying the original table that was passed to you. Didn't the tests break?

ElizabethSvit replied:
Yes, I removed it accidentially and tested only second algorithm after that.

ssancho2 commented on ass4.cpp:
```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
-        tile_deque.pop_front();
+    int find(int x) {
+        if (parent[x] != x) {
+            return find(parent[x]);
```
It's a very common optimization to set `parent[x] = find(parent[x])` this way the tree gets shallower and it's faster the next time around.

ssancho2 commented on ass4.cpp:
```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
+            if (table[y][x]) {
+                for (auto delta : delta_x_y) {
```
Actually, you only need to join with 2, up and left (or up and right or similar :) ), that may make the code simpler.

ssancho2 commented on ass4.cpp:
```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
+    }
+    
+    Disjoint_sets islands = Disjoint_sets(table.size() * max_width);
+    vector<pair<int, int>> delta_x_y = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
```
Nice

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,135 @@
+    int max_width = 0;
+    
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
```
No need for this loop.

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    vector<int> components(table.size() * max_width);
```
Easier to use a set, add all values returned by find on the islands and see the size at the end.

ssancho2 commented on ass4.cpp:
```diff
@@ -0,0 +1,135 @@
+
+int main(int argc, const char * argv[]) {
+    vector<vector<bool> > table = {{0,0,1}, {1,0,0}, {1,1,0}};
+    cout << count_islands(table) << endl;
```
Some more test will be nice :)

ssancho2 reviewed (approved):
Some comments, but fell free to move to the next assignment if you don't have time.
